File: particle_filter/image_dif.cpp
// about data type, (eg. what does CV_64S1 referes to )
//https://docs.opencv.org/2.4/modules/core/doc/basic_structures.html#vec
// find data type of mat in opencv
// https://codeyarns.com/2015/08/27/depth-and-type-of-matrix-in-opencv/

#include "image_dif.hpp"

cv::Vec3f bgr2hsi(cv::Vec3b bgr)
{
  float B = bgr.val[0]/255.0;
  float G = bgr.val[1]/255.0;
  float R = bgr.val[2]/255.0;
  float I = (R + G + B)/3.0;
  float S = 1.0 - std::min({B, G, R})/I;
  float H = acos(0.5*((R-G)+(R-B))/sqrt((R-G)*(R-G)+(R-B)*(G-B)));
  if(B>G){
    H = 2*3.1415 - H;
  }
  cv::Vec3f hsi(H, S, I);
  return hsi;
}

cv::Mat convert_bf(const cv::Mat& img, std::function<bool(cv::Vec3b)> predicate)
{
  cv::Mat img_bf(img.rows, img.cols, CV_8UC3); // CV_8UC3 really?
  for (int i = 0; i < img.rows; i++){
    for (int j = 0; j < img.cols; j++){
      auto bgr = img.at<cv::Vec3b>(i, j);
      auto bgr_white = cv::Vec3b(255, 255, 255);
      auto bgr_black = cv::Vec3b(0, 0, 0);
      img_bf.at<cv::Vec3b>(i, j) = predicate(bgr) ? bgr_white : bgr_black;
    }
  }
  return img_bf;
}

std::function<bool(cv::Vec3b)> gen_hsi_filter(
    float h_min, float h_max,
    float s_min, float s_max,
    float i_min, float i_max)
{
  auto predicate = [=](cv::Vec3b bgr){
    auto hsi = bgr2hsi(bgr);
    float h = hsi.val[0];
    float s = hsi.val[1];
    float i = hsi.val[2];
    if(h < h_min || h_max < h) {return false;}
    if(s < s_min || s_max < s) {return false;}
    if(i < i_min || i_max < i) {return false;}
    return true;
  };
  return predicate;
}

int compute_cost(cv::Mat img1, cv::Mat img2)
{
  int cost = 0;
  for (int i = 0; i < img1.rows; i++){
    for (int j = 0; j < img1.cols; j++){
      auto bgr1 = img1.at<cv::Vec3b>(i, j);
      auto bgr2 = img2.at<cv::Vec3b>(i, j);
      if (cv::sum(bgr1) != cv::sum(bgr2)){cost++;}
    }
  }
  return cost;
}

File: particle_filter/image_dif.py
import cv2
import math
import numpy as np
import matplotlib.pyplot as plt
inf = 10000000000000000

def bgr2hsi(bgr):
    b, g, r = bgr/255.0
    i = (r + g + b)/3.0
    s = 1.0 - min([b, g, r])/i
    h = math.acos(0.5*((r-g)+(r-b))/(math.sqrt((r-g)**2+(r-b)*(g-b))))
    if b<= g:
        h = h
    else:
        h = math.pi*2 - h
    hsi = np.array([h, s, i])
    return hsi

def image_to_pixel_list(img):
    dim_x = img.shape[0]
    dim_y = img.shape[1]
    pixel_list = [[i, j] for i in range(dim_x) for j in range(dim_y)]
    return pixel_list

def convert_bf(img, predicate):
    img_filled = np.copy(img)
    for pixel in image_to_pixel_list(img):
        rgb = img[pixel[0]][pixel[1]]
        if predicate(rgb):
            img_filled[pixel[0]][pixel[1]] = np.array([255, 255, 255])
        else:
            img_filled[pixel[0]][pixel[1]] = np.array([0, 0, 0])
    return img_filled

def gen_hsi_filter(h_bound, s_bound, i_bound):
    # ex: h_bound = [h_min, h_max]
    def hsi_filter(bgr):
        hsi = bgr2hsi(bgr)
        h = hsi[0]; s = hsi[1]; i = hsi[2]
        if h < h_bound[0] or h_bound[1] < h:
            return False
        if s < s_bound[0] or s_bound[1] < s:
            return False
        if i < i_bound[0] or i_bound[1] < i:
            return False
        return True
    return hsi_filter

def evaluatedif(img1, img2):
    img_dif = img1 - img2
    cost = 0
    for pixel in image_to_pixel_list(img_dif):
        rgb = img_dif[pixel[0]][pixel[1]]
        if sum(rgb)!=0:
            cost += 1
    return cost


img1 = cv2.imread("image/pre.png")
img2 = cv2.imread("image/post_with_move.png")
img1_ = convert_bf(img1, gen_hsi_filter([-inf, inf], [0.3, 1], [0.5, 0.8]))
img2_ = convert_bf(img2, gen_hsi_filter([-inf, inf], [0.3, 1], [0.5, 0.8]))
cost = evaluatedif(img1_, img2_)
#plt.imshow(img_dif)
#plt.show()

